File: Programmers/[64062] 징검다리 건너기/Johoseong/64062.py
def solution(stones, k):
    answer = 2000000000
    start = min(stones)
    end = max(stones)

    while start <= end:
        mid = (start + end) // 2

        count = 0
        for s in stones: # 연속된 못 밟는 돌 개수 셈
            if s - mid <= 0:
                count += 1
            else:
                count = 0

            if k <= count:
                break

        if k <= count: # 못 밟는 돌 개수가 k보다 큼 (징검다리 못 건너는 경우)
            end = mid - 1
            answer = min(answer, mid)
        else: # 징검다리 건널 수 있는 경우
            start = mid + 1

    return answer

print(solution([2, 4, 5, 3, 2, 1, 4, 2, 5, 1], 3))

File: Programmers/[64062] 징검다리 건너기/ONE/64062.java
class Solution {
    public int solution(int[] stones, int k) {
        int max = Integer.MIN_VALUE;
        int min = Integer.MAX_VALUE;

        for (int stone : stones) {
            max = Math.max(max, stone);
            min = Math.min(min, stone);
        }

        return parametricSearch(stones, k, min, max);
    }

    private int parametricSearch(int[] stones, int k, int low, int high) {
        if (low == high)
            return low;

        while (low < high) {
            int mid = (low + high) / 2;

            if (cross(stones, k, mid))
                low = mid + 1;
            else
                high = mid;
        }

        return low - 1;
    }

    private boolean cross(int[] stones, int k, int mid) {
        int cnt = 0;

        for (int stone : stones) {
            if (stone - mid < 0)
                cnt++;
            else
                cnt = 0;

            if (cnt == k)
                return false;
        }
        return true;
    }
}